File: java/socketd/src/main/java/org/noear/socketd/transport/core/ConfigBase.java
package org.noear.socketd.transport.core;

import org.noear.socketd.transport.core.buffer.BufferReader;
import org.noear.socketd.transport.core.buffer.BufferWriter;
import org.noear.socketd.transport.core.identifier.GuidGenerator;
import org.noear.socketd.transport.core.fragment.FragmentHandlerDefault;
import org.noear.socketd.transport.core.stream.StreamMangerDefault;
import org.noear.socketd.utils.NamedThreadFactory;

import javax.net.ssl.SSLContext;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.*;

/**
 * @author noear
 * @since 2.0
 */
public abstract class ConfigBase<T extends Config> implements Config {
    //是否客户端模式
    private final boolean clientMode;
    //流管理器
    private final StreamManger streamManger;

    //字符集
    protected Charset charset;

    //编解码器
    protected Codec<BufferReader, BufferWriter> codec;
    //id生成器
    protected IdGenerator idGenerator;
    //分片处理
    protected FragmentHandler fragmentHandler;

    //ssl 上下文
    protected SSLContext sslContext;
    //通道执行器
    protected ExecutorService channelExecutor;

    //内核线程数
    protected int coreThreads;
    //最大线程数
    protected int maxThreads;

    //读缓冲大小
    protected int readBufferSize;
    //写缓冲大小
    protected int writeBufferSize;

    //连接空闲超时
    protected long idleTimeout;
    //请求默认超时
    protected long requestTimeout;
    //消息流超时（从发起到应答结束）
    protected long streamTimeout;
    //最大udp包大小
    protected int maxUdpSize;

    public ConfigBase(boolean clientMode) {
        this.clientMode = clientMode;
        this.streamManger = new StreamMangerDefault(this);

        this.charset = StandardCharsets.UTF_8;

        this.codec = new CodecByteBuffer(this);
        this.idGenerator = new GuidGenerator();
        this.fragmentHandler = new FragmentHandlerDefault();

        this.coreThreads = Math.max(Runtime.getRuntime().availableProcessors(), 2);
        this.maxThreads = coreThreads * 4;

        this.readBufferSize = 512;
        this.writeBufferSize = 512;

        this.idleTimeout = 0L; //默认不关（提供用户特殊场景选择）
        this.requestTimeout = 10_000L; //10秒（默认与连接超时同）
        this.streamTimeout = 1000 * 60 * 60 * 2;//2小时 //避免永不回调时，不能释放
        this.maxUdpSize = 2048; //2k //与 netty 保持一致 //实际可用 1464
    }

    /**
     * 是否客户端模式
     */
    @Override
    public boolean clientMode() {
        return clientMode;
    }

    @Override
    public StreamManger getStreamManger() {
        return streamManger;
    }

    /**
     * 获取角色名
     * */
    @Override
    public String getRoleName() {
        return clientMode() ? "Client" : "Server";
    }

    /**
     * 获取字符集
     */
    @Override
    public Charset getCharset() {
        return charset;
    }

    /**
     * 配置字符集
     */
    public T charset(Charset charset) {
        this.charset = charset;
        return (T) this;
    }

    /**
     * 获取编解码器
     */
    @Override
    public Codec<BufferReader, BufferWriter> getCodec() {
        return codec;
    }

    /**
     * 配置编解码器
     */
    public T codec(Codec<BufferReader, BufferWriter> codec) {
        Asserts.assertNull("codec", codec);

        this.codec = codec;
        return (T) this;
    }

    /**
     * 获取分片处理
     */
    @Override
    public FragmentHandler getFragmentHandler() {
        return fragmentHandler;
    }

    /**
     * 配置分片处理
     */
    public T fragmentHandler(FragmentHandler fragmentHandler) {
        Asserts.assertNull("fragmentHandler", fragmentHandler);

        this.fragmentHandler = fragmentHandler;
        return (T) this;
    }

    /**
     * 获取标识生成器
     */
    @Override
    public IdGenerator getIdGenerator() {
        return idGenerator;
    }

    /**
     * 配置标识生成器
     */
    public T idGenerator(IdGenerator idGenerator) {
        Asserts.assertNull("idGenerator", idGenerator);

        this.idGenerator = idGenerator;
        return (T) this;
    }

    /**
     * 获取 ssl 上下文
     */
    @Override
    public SSLContext getSslContext() {
        return sslContext;
    }

    /**
     * 配置 ssl 上下文
     */
    public T sslContext(SSLContext sslContext) {
        this.sslContext = sslContext;
        return (T) this;
    }

    private Object EXECUTOR_LOCK = new Object();

    @Override
    public ExecutorService getChannelExecutor() {
        if (channelExecutor == null) {
            synchronized (EXECUTOR_LOCK) {
                if (channelExecutor == null) {
                    int nThreads = clientMode() ? coreThreads : maxThreads;

                    channelExecutor = new ThreadPoolExecutor(nThreads, nThreads,
                            0L, TimeUnit.MILLISECONDS,
                            new LinkedBlockingQueue<Runnable>(),
                            new NamedThreadFactory("Socketd-channelExecutor-"));
                }
            }
        }

        return channelExecutor;
    }

    /**
     * 配置调试执行器
     * */
    public T channelExecutor(ExecutorService channelExecutor) {
        ExecutorService odl = this.channelExecutor;
        this.channelExecutor = channelExecutor;

        if (odl != null) {
            odl.shutdown();
        }

        return (T) this;
    }

    /**
     * 获取核心线程数
     */
    @Override
    public int getCoreThreads() {
        return coreThreads;
    }

    /**
     * 配置核心线程数
     */
    public T coreThreads(int coreThreads) {
        this.coreThreads = coreThreads;
        return (T) this;
    }

    /**
     * 获取最大线程数
     */
    @Override
    public int getMaxThreads() {
        return maxThreads;
    }

    /**
     * 配置最大线程数
     */
    public T maxThreads(int maxThreads) {
        this.maxThreads = maxThreads;
        return (T) this;
    }


    /**
     * 获取读缓冲大小
     */
    public int getReadBufferSize() {
        return readBufferSize;
    }

    /**
     * 配置读缓冲大小
     */
    public T readBufferSize(int readBufferSize) {
        this.readBufferSize = readBufferSize;
        return (T) this;
    }

    /**
     * 获取写缓冲大小
     */
    public int getWriteBufferSize() {
        return writeBufferSize;
    }

    /**
     * 配置写缓冲大小
     */
    public T writeBufferSize(int writeBufferSize) {
        this.writeBufferSize = writeBufferSize;
        return (T) this;
    }


    /**
     * 获取连接空闲超时
     */
    public long getIdleTimeout() {
        return idleTimeout;
    }

    /**
     * 配置连接空闲超时
     */
    public T idleTimeout(int idleTimeout) {
        this.idleTimeout = idleTimeout;
        return (T) this;
    }

    /**
     * 获取请求默认超时
     */
    @Override
    public long getRequestTimeout() {
        return requestTimeout;
    }

    /**
     * 配置请求默认超时
     */
    public T requestTimeout(long requestTimeout) {
        this.requestTimeout = requestTimeout;
        return (T) this;
    }

    /**
     * 获取消息流超时（单位：毫秒）
     * */
    @Override
    public long getStreamTimeout() {
        return streamTimeout;
    }

    /**
     * 配置消息流超时（单位：毫秒）
     * */
    public T streamTimeout(long streamTimeout) {
        this.streamTimeout = streamTimeout;
        return (T) this;
    }

    /**
     * 获取允许最大UDP包大小
     */
    @Override
    public int getMaxUdpSize() {
        return maxUdpSize;
    }

    /**
     * 配置允许最大UDP包大小
     */
    public T maxUdpSize(int maxUdpSize) {
        this.maxUdpSize = maxUdpSize;
        return (T) this;
    }
}

File: python/socketd/socketd/core/config/ConfigBase.py
import os
import ssl

from typing import Callable
from concurrent.futures import ThreadPoolExecutor

from .Config import Config
from socketd.core.handler.FragmentHandlerDefault import FragmentHandlerDefault
from socketd.transport.Codec import Codec
from socketd.transport.CodecByteBuffer import CodecByteBuffer


class ConfigBase(Config):

    def __init__(self, client_mode: bool):
        self._stream_timeout = 1000 * 60 * 60 * 2
        self._request_timeout = 10_000
        self._client_mode = client_mode
        self._charset = "utf-8"
        self._codec: Codec = CodecByteBuffer(self)
        self._idGenerator: Callable = None
        self._fragmentHandler: FragmentHandlerDefault = FragmentHandlerDefault()
        self._sslContext = None
        self._executor = None
        self._core_threads = os.cpu_count() * 2
        self._max_threads = self._core_threads * 8
        self._reply_timeout = 3000
        self._max_requests = 10
        self._maxUdpSize = 2048

    def client_mode(self):
        return self._client_mode

    def get_charset(self):
        return self._charset

    def charset(self, charset):
        self._charset = charset
        return self

    def get_codec(self):
        return self._codec

    def codec(self, codec):
        assert codec is None
        self._codec = codec
        return self

    def get_fragment_handler(self):
        return self._fragmentHandler

    def fragment_handler(self, fragmentHandler):
        assert fragmentHandler is None
        self._fragmentHandler = fragmentHandler
        return self

    def get_id_generator(self):
        return self._idGenerator

    def id_generator(self, _idGenerator):
        # assert _idGenerator is None
        self._idGenerator = _idGenerator
        return self

    def get_ssl_context(self):
        return self._sslContext

    def ssl_context(self, localhost_pem: str):
        self._sslContext = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
        self._sslContext.load_cert_chain(localhost_pem)
        return self

    def get_executor(self):
        if self._executor is None:
            __nThreads = self._core_threads if self.client_mode() else self._max_threads
            self._executor = ThreadPoolExecutor(max_workers=__nThreads, thread_name_prefix="socketd-channelExecutor-")
        return self._executor

    def set_executor(self, executor):
        self._executor = executor
        return self

    def get_core_threads(self):
        return self._core_threads

    def set_core_threads(self, core_threads):
        self._core_threads = core_threads
        return self

    def get_max_threads(self):
        return self._max_threads

    def set_max_threads(self, max_threads):
        self._max_threads = max_threads
        return self

    def get_reply_timeout(self):
        return self._reply_timeout

    def set_reply_timeout(self, reply_timeout):
        self._reply_timeout = reply_timeout
        return self

    def get_max_requests(self):
        return self._max_requests

    def set_max_requests(self, max_requests):
        self._max_requests = max_requests
        return self

    def get_max_udp_size(self):
        return self._maxUdpSize

    def set_max_udp_size(self, maxUdpSize):
        self._maxUdpSize = maxUdpSize
        return self

    def get_request_timeout(self) -> float:
        return self._request_timeout

    def set_request_timeout(self, _request_time_out):
        self._request_timeout = _request_time_out
        return self

    def get_stream_timeout(self) -> float:
        return self._stream_timeout

    def set_stream_timeout(self, _stream_timeout):
        self._stream_timeout = _stream_timeout
        return self
